[FLINK-20812][Connector][Hbase] hbase in sql mode,can use 'properties.*' add Configuration parameter.

chaozwn opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request makes hbase in sql mode,can use 'properties.*' add Configuration parameter
e.g 
      'properties.hbase.security.authentication'='kerberos',
      'properties.hbase.master.kerberos.principal'='...',

## Brief change log

Without affecting the original function, perfect access, custom Configuration parameters

## Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers:  no 
  - The runtime per-record code paths (performance sensitive):  no 
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no 
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  yes
  - If yes, how is the feature documented? not documented

wuchong commented:
cc @leonardBang for reviewing.

chaozwn commented:
> ## CI report:
> * [f66faaa](https://github.com/apache/flink/commit/f66faaa7fd4901814134ab3d8993c52838fd77bf) Azure: [FAILURE](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=11540)
> 
> Bot commands
> The @flinkbot bot supports the following commands:
> * `@flinkbot run travis` re-run the last Travis build
> * `@flinkbot run azure` re-run the last Azure build

@flinkbot run azure

chaozwn commented:
> * run azure

run azure

chaozwn reviewed (commented):
resolve NameTest

chaozwn commented:
@flinkbot run azure

chaozwn commented:
@flinkbot run azure

chaozwn commented:
@flinkbot run azure

chaozwn commented:
@flinkbot run azure

chaozwn commented:
@flinkbot run azure

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1TableFactory.java:
```diff
@@ -50,15 +50,17 @@
+import static org.apache.flink.connector.hbase1.HBaseValidator.CONNECTOR_WRITE_BUFFER_FLUSH_MAX_SIZE;
+import static org.apache.flink.connector.hbase1.HBaseValidator.CONNECTOR_ZK_NODE_PARENT;
+import static org.apache.flink.connector.hbase1.HBaseValidator.CONNECTOR_ZK_QUORUM;
+import static org.apache.flink.table.descriptors.AbstractHBaseValidator.CONNECTOR_PROPERTIES;
```
I think current import order is fine, if you want to optimize the imports please keep them in same style i.e.:
```suggestion
import static org.apache.flink.connector.hbase1.HBaseValidator.CONNECTOR_PROPERTIES.CONNECTOR_PROPERTIES;
```

leonardBang reviewed (changes requested):
@chaozwn Thanks for the contribution, the PR generally looks good, I left some comments about details.

chaozwn commented on flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/source/AbstractHBaseDynamicTableSource.java:
```diff
@@ -114,4 +117,29 @@ public String asSummaryString() {
     }
+
+    // get hbase table properties which start with prefix
+    public static Properties getSpecifyProperties(Map<String, String> tableOptions, String prefix) {
```
ok

chaozwn replied:
this is a method which we can get Specify properties from table options，eg getSpecifyProperties(options,"properties") or getSpecifyProperties(options,"properties.connector"),  More functions。

chaozwn replied:
now,all methods are changed, getHBaseClientProperties and containsHBaseClientProperties

chaozwn reviewed (commented):
ok

chaozwn commented on flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarSubmissionITCase.java:
```diff
@@ -95,7 +95,7 @@ public void testJarSubmission() throws Exception {
         // we're only interested in the core functionality so checking for a small detail is
         // sufficient
-        Assert.assertThat(planResponse.getJsonPlan(), containsString("TestProgram.java:30"));
+        Assert.assertThat(planResponse.getJsonPlan(), containsString("TestProgram.java:28"));
```
when i use master branch, this method test failed in flink bot ,because TestProgram change line

chaozwn replied:
Do I need to cancel the changes in these two test files, NameTest.java and TestProgram.java

chaozwn replied:
rollback

leonardBang replied:
I didn't encounter this problem, maybe you need rebase the latest master branch, but we shouldn't modify it in this PR.

chaozwn commented on flink-java/src/test/java/org/apache/flink/api/java/operators/NamesTest.java:
```diff
@@ -62,7 +62,7 @@ public boolean filter(String value) throws Exception {
                         })
                 .output(new DiscardingOutputFormat<String>());
         Plan plan = env.createProgramPlan();
-        testForName("Filter at testDefaultName(NamesTest.java:55)", plan);
```
ok

chaozwn replied:
rollback

chaozwn reviewed (commented):
Call back 2 test files

chaozwn reviewed (commented):
Call back 2 test files

leonardBang commented:
Hint：the PR is not a large one，you can squash these commits to one commit and then edit the files ^_^

chaozwn commented:
> Hint：the PR is not a large one，you can squash these commits to one commit and then edit the files ^_^

ok，Do I need to reopen a pr?

chaozwn commented:
> Hint：the PR is not a large one，you can squash these commits to one commit and then edit the files ^_^

Modification completed

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1DynamicTableFactory.java:
```diff
@@ -104,10 +106,14 @@
                                     + "Can be set to '0' to disable it. Note, both 'sink.buffer-flush.max-size' and 'sink.buffer-flush.max-rows' "
                                     + "can be set to '0' with the flush interval set allowing for complete async processing of buffered actions.");
 
+    // Prefix for Hbase specific properties.
```
```suggestion
    // Prefix for HBase specific properties.
```

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1DynamicTableFactory.java:
```diff
@@ -118,6 +124,10 @@ public DynamicTableSource createDynamicTableSource(Context context) {
         hbaseClientConf.set(HConstants.ZOOKEEPER_QUORUM, helper.getOptions().get(ZOOKEEPER_QUORUM));
         hbaseClientConf.set(
                 HConstants.ZOOKEEPER_ZNODE_PARENT, helper.getOptions().get(ZOOKEEPER_ZNODE_PARENT));
+        // add hbase properties
```
```suggestion
        // add HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1TableFactory.java:
```diff
@@ -88,7 +90,14 @@
                 .getOptionalString(CONNECTOR_ZK_NODE_PARENT)
                 .ifPresent(v -> hbaseClientConf.set(HConstants.ZOOKEEPER_ZNODE_PARENT, v));
-
+        // add hbase properties
```
```suggestion
        // add HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1TableFactory.java:
```diff
@@ -106,6 +115,16 @@
                 .getOptionalString(CONNECTOR_ZK_NODE_PARENT)
                 .ifPresent(hbaseOptionsBuilder::setZkNodeParent);
+        Properties hbaseProperties = new Properties();
+        // add hbase properties
```
```suggestion
        // add HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1TableFactory.java:
```diff
@@ -200,6 +219,9 @@ private DescriptorProperties getValidatedProperties(Map<String, String> properti
         properties.add(SCHEMA + "." + DescriptorProperties.PRIMARY_KEY_NAME);
         properties.add(SCHEMA + "." + DescriptorProperties.PRIMARY_KEY_COLUMNS);
 
+        // hbase properties
```
```suggestion
        // HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/sink/HBaseUpsertTableSink.java:
```diff
@@ -96,6 +96,13 @@ public TableSchema getTableSchema() {
         hbaseOptions
                 .getZkNodeParent()
                 .ifPresent(v -> hbaseClientConf.set(HConstants.ZOOKEEPER_ZNODE_PARENT, v));
+        // add hbase properties
```
```suggestion
        // add HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-2.2/src/main/java/org/apache/flink/connector/hbase2/HBase2DynamicTableFactory.java:
```diff
@@ -103,31 +105,46 @@
                                     + "Can be set to '0' to disable it. Note, both 'sink.buffer-flush.max-size' and 'sink.buffer-flush.max-rows' "
                                     + "can be set to '0' with the flush interval set allowing for complete async processing of buffered actions.");
 
+    // Prefix for Hbase specific properties.
```
```suggestion
    // Prefix for HBase specific properties.
```

leonardBang commented on flink-connectors/flink-connector-hbase-2.2/src/main/java/org/apache/flink/connector/hbase2/HBase2DynamicTableFactory.java:
```diff
@@ -103,31 +105,46 @@
         String hTableName = helper.getOptions().get(TABLE_NAME);
 
+        Configuration hbaseConf = getHbaseConf(helper);
+        // add hbase properties
```
```suggestion
        // add HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-2.2/src/main/java/org/apache/flink/connector/hbase2/HBase2TableFactory.java:
```diff
@@ -80,6 +81,15 @@
         final DescriptorProperties descriptorProperties = getValidatedProperties(properties);
         Configuration hbaseClientConf = getHConf(descriptorProperties);
 
+        // add hbase properties
```
```suggestion
        // add HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-2.2/src/main/java/org/apache/flink/connector/hbase2/HBase2TableFactory.java:
```diff
@@ -96,6 +106,16 @@
         HBaseTableSchema hbaseSchema = validateTableSchema(tableSchema);
 
+        Configuration hbaseClientConf = getHConf(descriptorProperties);
+        // add hbase properties
```
```suggestion
        // add HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-2.2/src/main/java/org/apache/flink/connector/hbase2/HBase2TableFactory.java:
```diff
@@ -188,6 +208,9 @@ private DescriptorProperties getValidatedProperties(Map<String, String> properti
         properties.add(SCHEMA + "." + DescriptorProperties.PRIMARY_KEY_NAME);
         properties.add(SCHEMA + "." + DescriptorProperties.PRIMARY_KEY_COLUMNS);
 
+        // hbase properties
```
```suggestion
        // HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/source/AbstractHBaseDynamicTableSource.java:
```diff
@@ -114,4 +117,29 @@ public String asSummaryString() {
         return this.hbaseSchema;
     }
+
+    // get hbase table properties which start with prefix
```
```suggestion
    // get HBase table properties which start with prefix
```

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1DynamicTableFactory.java:
```diff
@@ -128,14 +138,20 @@ public DynamicTableSource createDynamicTableSource(Context context) {
         TableSchema tableSchema = context.getCatalogTable().getSchema();
         validatePrimaryKey(tableSchema);
 
+        // add hbase properties
```
```suggestion
        // add HBase properties
```

leonardBang commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/options/HBaseOptions.java:
```diff
@@ -120,11 +136,18 @@ public Builder setZkNodeParent(String zkNodeParent) {
             return this;
         }
 
+        /** Optional. Sets hbase properties for hbase configuration. */
```
```suggestion
        /** Optional. Sets HBase properties for hbase configuration. */
```

leonardBang reviewed (commented):
Thanks for the update, I only left some typo comments, could you rebase master and squash to one commit ?

chaozwn commented:
> Thanks for the update, I only left some typo comments, could you rebase master and squash to one commit ?

ok

leonardBang reviewed (approved):
LGTM CC @wuchong

wuchong commented on flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/source/AbstractHBaseDynamicTableSource.java:
```diff
@@ -114,4 +117,28 @@ public String asSummaryString() {
+
+    // get HBase table properties which start with prefix
+    public static Properties getHBaseClientProperties(
+            Map<String, String> tableOptions, String prefix) {
```
I think we don't need the `prefix` parameter, the prefix should always be `properties.`.

chaozwn replied:
yes,i will modify

wuchong reviewed (commented):
Thanks for the great contribution @chaozwn . The code looks good overall. I would just want to improve the implementation a bit more. 

What confuses me a lot is that the implementaion for source and sink is different. For the source, we have merged the HBase properties into Hadoop Configuration before constructing the table source. However, for the sink, we pass the properties via `HBaseOptions` and merge the properties into Hadoop Configuration in `getSinkRuntimeProvider`. 

I would suggest to unify the implementation to Hadoop Configuration for better future evolution. We can even get rid of `HBaseOptions` then. The benefit is that we can have an `HBaseOptions` class in the base module and put HBase connector options and utilities (e.g. `getHBaseClientProperties`) into it.

What do you think?

wuchong commented on flink-connectors/flink-connector-hbase-1.4/src/test/java/org/apache/flink/connector/hbase1/HBaseDynamicTableFactoryTest.java:
```diff
@@ -183,18 +181,6 @@ public void testTableSinkFactory() {
-                        .setHbaseProperties(properties)
-                        .build();
-        HBaseOptions actualHBaseOptions = hbaseSink.getHBaseOptions();
-        assertEquals(expectedHBaseOptions, actualHBaseOptions);
```
We should add another way to test the configurations. For example : we can get the Hadoop Configruation and convert it into an Map via `Configruation.iterator()`, the we can use `assertEquals` to verify the configs. 

We should also remember to verify table name separately, because it is not in the Hadoop Configruation now.

wuchong reviewed (changes requested):
Thanks for the updating. I left some comments.

wuchong commented on flink-connectors/flink-connector-hbase-1.4/src/test/java/org/apache/flink/connector/hbase1/HBaseDynamicTableFactoryTest.java:
```diff
@@ -181,6 +185,25 @@ public void testTableSinkFactory() {
                 hbaseSchema.getQualifierDataTypes("f4"));
 
+        // verify hadoop Configuration
+        org.apache.hadoop.conf.Configuration expectedConfiguration =
+                HBaseConfigurationUtil.getHBaseConfiguration();
+        expectedConfiguration.set(HConstants.ZOOKEEPER_QUORUM, "localhost:2181");
+        expectedConfiguration.set(HConstants.ZOOKEEPER_ZNODE_PARENT, "/flink");
+        expectedConfiguration.set("hbase.security.authentication", "kerberos");
+        Map<String, String> expectedProperties =
+                Lists.newArrayList(expectedConfiguration.iterator()).stream()
+                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
We can directly construct the expected Map.

chaozwn replied:
solved

wuchong commented on flink-connectors/flink-connector-hbase-2.2/src/test/java/org/apache/flink/connector/hbase2/HBaseDynamicTableFactoryTest.java:
```diff
@@ -181,6 +185,25 @@ public void testTableSinkFactory() {
+        expectedConfiguration.set(HConstants.ZOOKEEPER_QUORUM, "localhost:2181");
+        expectedConfiguration.set(HConstants.ZOOKEEPER_ZNODE_PARENT, "/flink");
+        expectedConfiguration.set("hbase.security.authentication", "kerberos");
+        Map<String, String> expectedProperties =
```
ditto.

chaozwn replied:
solved

wuchong commented on flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1DynamicTableFactory.java:
```diff
@@ -74,13 +72,13 @@ public DynamicTableSource createDynamicTableSource(Context context) {
         validatePrimaryKey(tableSchema);
         validateTableSourceOptions(tableOptions);
```
We don't need this, this has been checked in `helper.validateExcept(PROPERTIES_PREFIX)`.

chaozwn replied:
validatePrimaryKey is necessary，The other can be removed.

wuchong reviewed (approved):
LGTM.

wuchong commented:
Thanks for the contribution @chaozwn , I rebased and squashed the pull request. Will merge it once build is passed. 

I found that the documentation is missed, could you create an issue to add docs and would be great if you can take it too.

chaozwn commented:
> Thanks for the contribution @chaozwn , I rebased and squashed the pull request. Will merge it once build is passed.
> 
> I found that the documentation is missed? Could you create an issue to add docs? And would be great if you can take it too.

ok, i will do it.

wuchong commented:
Build is passed: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=12029&view=results
Merging...
